### ISSUE TITLE
Prometheus metric ﻿bucket_objects_count filter

### ISSUE THREAD

**Issue by buhl** [minio/minio#10568]
**Is your feature request related to a problem? Please describe.**
I have buckets being used to send files out to processing jobs out of my control.
The agreement is that the processing job can delete a file once it has been processed.
I keep the files in these buckets versioned so that I can quickly restore a wrongfully deleted file.
I would like to use the ﻿`bucket_objects_count` to measure if the buckets are being drained in a timely manner by the remote processors so that I can react to any potential errors.
My problem is that ﻿`bucket_objects_count` gives the total count of files in a bucket including the deleted ones. So I have no way of telling if files are being deleted

**Describe the solution you'd like**
I would like to have a filter `object_version` on `﻿bucket_objects_count` similar to `bucket_objects_histogram﻿`'s `object_size`.
`object_version` should be one of `previous`, `current`, `removed`. Where `previous` would be older versions of a `current` or `removed`. `current` should a the number of latest version files. and `removed` should be number of latest version files that have been removed.
I am a little unsure if the `removed` marker makes sense since you can upload a file, delete it and upload it again and it will have 3 versions where the middle one will have a deleted-marker on it.
But what I need most of all is a count of all the latest versions of files without a deleted-marker.

**Describe alternatives you've considered**
Right now I have scripts monitoring these buckets feeding the information into prometheus.
I intertained the idea of suggesting to have a bucket label on ﻿s3_requests_current but I ended up with suggestion the filter on ﻿bucket_objects_count as it seems more appropriate

**Comment by buhl** [comment]
On a semi side note. `﻿bucket_objects_count` is a count of all files including deleted files but not including all versions of a file.
I don't know if this is on purpose, but it does seem a bit misleading?

**Comment by harshavardhana** [comment]
> On a semi side note. `﻿bucket_objects_count` is a count of all files including deleted files but not including all versions of a file.
> I don't know if this is on purpose, but it does seem a bit misleading?

bucket_objects_count is all files including deleted, as well as including all versions.

**Comment by buhl** [comment]
It might just be me misunderstanding but `bucket_objects_count` reports 10
```
bucket_objects_count{bucket="unknown-source",instance="minio1:9000",job="minio"} | 10
```
I have 9 files
```
mc ls minio/count-test
[2020-09-25 10:39:53 UTC]      5B 3ae463d6-08ab-4d7b-85f5-77f046696bc5
[2020-09-25 10:21:33 UTC]      3B 5aab27cf-aeb3-4a1f-97bd-272ba67e98ff
[2020-09-25 10:11:21 UTC]      2B 7aa72eb8-8dfc-4d9f-aa11-0065b14ff500
[2020-09-25 10:21:33 UTC]      4B 84647fca-3a0b-45d1-a1b9-16b6b71939d4
[2020-09-25 10:11:20 UTC]      2B a30b103f-86f2-4a3d-abb3-c41438dc177f
[2020-09-25 10:11:20 UTC]      2B e9f5560f-c2df-4944-846c-82eca37035e2
[2020-09-25 10:11:20 UTC]      2B f28e34f2-004b-40c0-8d77-8b224071f140
[2020-09-25 10:39:53 UTC]      2B fb20f70a-654b-428c-97db-ef2b37cb5d36
[2020-09-25 10:21:33 UTC]      4B fc5c5aa5-1b6a-4803-811a-eef232dfab1a
```
and 25 versions.
```
mc ls --versions minio/count-test
[2020-09-25 10:39:53 UTC]      5B 3ae463d6-08ab-4d7b-85f5-77f046696bc5 (5th, vid=edebed66-3786-4426-9d8e-5009b815bb8d)
[2020-09-25 10:27:17 UTC]      0B 3ae463d6-08ab-4d7b-85f5-77f046696bc5 (4th, vid=5c9f9a13-9b62-4872-bd13-49ae09c9d6a3, delete-marker)
[2020-09-25 10:21:33 UTC]      5B 3ae463d6-08ab-4d7b-85f5-77f046696bc5 (3rd, vid=9b436d5f-963d-494c-98ea-a870f1614785)
[2020-09-25 10:15:54 UTC]      3B 3ae463d6-08ab-4d7b-85f5-77f046696bc5 (2nd, vid=90e00643-c976-42c7-beee-66bd3aaec662)
[2020-09-25 10:11:20 UTC]      2B 3ae463d6-08ab-4d7b-85f5-77f046696bc5 (1st, vid=e585da42-3299-4483-84af-9427f6b2f2af)
[2020-09-25 10:21:33 UTC]      3B 5aab27cf-aeb3-4a1f-97bd-272ba67e98ff (3rd, vid=97be5814-3ad4-4101-af26-51b2acfdd600)
[2020-09-25 10:15:54 UTC]      2B 5aab27cf-aeb3-4a1f-97bd-272ba67e98ff (2nd, vid=05736ef8-8932-4881-b1a6-1f73692aa7ed)
[2020-09-25 10:11:21 UTC]      2B 5aab27cf-aeb3-4a1f-97bd-272ba67e98ff (1st, vid=6e781ce5-bf86-46e4-a7f3-647a2485660d)
[2020-09-25 10:11:21 UTC]      2B 7aa72eb8-8dfc-4d9f-aa11-0065b14ff500 (1st, vid=0f61888f-cd06-4148-853e-d4288aae8eca)
[2020-09-25 10:21:33 UTC]      4B 84647fca-3a0b-45d1-a1b9-16b6b71939d4 (3rd, vid=9741e4b7-8e1c-4cba-a27a-8640f42ceada)
[2020-09-25 10:15:54 UTC]      3B 84647fca-3a0b-45d1-a1b9-16b6b71939d4 (2nd, vid=3ecf87bd-4bc4-4179-9a90-2672c1e4cfd2)
[2020-09-25 10:11:20 UTC]      2B 84647fca-3a0b-45d1-a1b9-16b6b71939d4 (1st, vid=3bda1505-126a-469b-9751-b56cf2d1a0ff)
[2020-10-03 17:52:11 UTC]      0B 9fa9a14c-7a64-434c-b690-a17369ca08fd (4th, vid=7b796f12-0795-4300-9f1b-2b6d249c8238, delete-marker)
[2020-09-25 10:21:33 UTC]      5B 9fa9a14c-7a64-434c-b690-a17369ca08fd (3rd, vid=673ea3c1-07c4-4b46-bfe0-d606797928a1)
[2020-09-25 10:15:54 UTC]      3B 9fa9a14c-7a64-434c-b690-a17369ca08fd (2nd, vid=a237a74e-92e9-4fa5-aa19-005d5516ed24)
[2020-09-25 10:11:20 UTC]      2B 9fa9a14c-7a64-434c-b690-a17369ca08fd (1st, vid=37f1b8a8-194d-4a61-9475-8dab6108058b)
[2020-09-25 10:11:20 UTC]      2B a30b103f-86f2-4a3d-abb3-c41438dc177f (1st, vid=d86cb208-3d2c-43b3-aed3-6f1f828d8b76)
[2020-09-25 10:11:20 UTC]      2B e9f5560f-c2df-4944-846c-82eca37035e2 (1st, vid=b71be97c-62e6-4338-acb7-ff2f5ccb61ba)
[2020-09-25 10:11:20 UTC]      2B f28e34f2-004b-40c0-8d77-8b224071f140 (1st, vid=50210efa-a42d-438a-aa67-28390cb243d8)
[2020-09-25 10:39:53 UTC]      2B fb20f70a-654b-428c-97db-ef2b37cb5d36 (3rd, vid=eb6f7413-eb0e-45df-8272-8424a1149baa)
[2020-09-25 10:31:25 UTC]      0B fb20f70a-654b-428c-97db-ef2b37cb5d36 (2nd, vid=4f0e00bd-160f-4c98-9c11-7dad06d0d076, delete-marker)
[2020-09-25 10:11:21 UTC]      2B fb20f70a-654b-428c-97db-ef2b37cb5d36 (1st, vid=d1577dc0-8c31-4a0a-be4c-f9446bac3152)
[2020-09-25 10:21:33 UTC]      4B fc5c5aa5-1b6a-4803-811a-eef232dfab1a (3rd, vid=3f3f96be-5b74-4754-a0f1-c8a02d1d174a)
[2020-09-25 10:15:54 UTC]      3B fc5c5aa5-1b6a-4803-811a-eef232dfab1a (2nd, vid=7073505b-4d6d-4763-81ca-173ba44df573)
[2020-09-25 10:11:20 UTC]      2B fc5c5aa5-1b6a-4803-811a-eef232dfab1a (1st, vid=fba6be42-f35f-47b5-afd9-34ad09f65263)
```
granted I _had_ 10 files at some point. but now I have 9 or a total of 25 version.

**Comment by vadmeste** [comment]
@buhl yeah correct, bucket_objects_count only returns the number of objects and not versions.

> I would like to use the ﻿`bucket_objects_count` to measure if the buckets are being drained in a timely manner by the remote processors so that I can react to any potential errors.

I think you can simply use bucket_data_usage in this case, because it returns the total size of all versions

**Comment by buhl** [comment]
@vadmeste  I have a use case where a remote processing job works on files in a bucket. Once the file is processed the remote job deletes it. I keep deleted versions in the bucket as a form for backup if the remote processing job has failed some how.
I am not in control of the remote processing job.
So i would like to have a metric for how many none deleted files there is left in a bucket. that way I can keep an eye on whether or not the remote job is keeping up with the amount of files in the bucket so I can stop my end from writing files to the bucket if the remote job is slow or stopped.
I hope this makes sense and I hope it's something you would want to implement.

Ideally I would like some kind of histogram

bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="current"} 9
bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="delete-marker"} 3
bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="1st"} 10
bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="2nd"} 6
bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="3rd"} 6
bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="4th"} 2
bucket_objects_hist{bucket="count-test", instance="minio1:9000",job="minio", version="5th"} 1

or something like this that can tell me how many none deleted files are in the bucket

**Comment by buhl** [comment]
I'd like to keep this open. as mentioned in https://github.com/minio/minio/issues/10568#issuecomment-703141858 I still think bucket_objects_count is misleading. It should be either 9 or 25, not 10.

**Comment by buhl** [comment]
I am still interested in having this resolved. I have no go skills, but I'd like to help in any other way I can.

**Comment by buhl** [comment]
I am okay with a low priority :+1:

**Comment by harshavardhana** [comment]
Fixed in #11141
